wazero/instance.go

```diff
@@ -0,0 +1,315 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package wazero
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/tetratelabs/wazero/api"
+	"github.com/tetratelabs/wazero/imports/wasi_snapshot_preview1"
+	"sync"
+	"sync/atomic"
+
+	wazero "github.com/tetratelabs/wazero"
+	"mosn.io/proxy-wasm-go-host/proxywasm/common"
+)
+
+var (
+	ErrInstanceNotStart     = errors.New("instance has not started")
+	ErrInstanceAlreadyStart = errors.New("instance has already started")
+)
+
+type Instance struct {
+	vm      *VM
+	module  *Module
+	imports map[string]wazero.HostModuleBuilder
+
+	instance api.Module
+
+	lock     sync.Mutex
+	started  uint32
+	refCount int
+	stopCond *sync.Cond
+
+	// user-defined data
+	data interface{}
+}
+
+type InstanceOptions func(instance *Instance)
+
+func NewInstance(vm *VM, module *Module, options ...InstanceOptions) *Instance {
+	r := vm.engine
+	ctx := context.Background()
+
+	ins := &Instance{
+		vm:      vm,
+		imports: map[string]wazero.HostModuleBuilder{},
+		module:  module,
+		lock:    sync.Mutex{},
+	}
+	ins.stopCond = sync.NewCond(&ins.lock)
+
+	for _, option := range options {
+		option(ins)
+	}
+
+	if _, err := wasi_snapshot_preview1.Instantiate(ctx, r); err != nil {
+		_ = r.Close(ctx)
+		panic(err)
+	}
+
+	return ins
+}
+
+func (w *Instance) GetData() interface{} {
+	return w.data
+}
+
+func (w *Instance) SetData(data interface{}) {
+	w.data = data
+}
+
+func (w *Instance) Acquire() bool {
```

Comment:
I kept all the locking as-is except there's no need to cache function or memory references in wazero.